### DISCUSSION TITLE
Feature Request: Specify number of files per partition for partitioned writes

### DISCUSSION THREAD

**Discussion by Hudler** [duckdb/duckdb#10346] [Ideas]
DuckDB allows creating hives via [partitioned writes](https://duckdb.org/docs/data/partitioning/partitioned_writes.html). Currently, the number of files per single partition is equal to the number of threads. 

If a user wants a single file per partition, he/she can limit the number of threads to 1. This however limits the performance greatly. It would be great if the partitioned writes could utilize more threads and still produce single file per partition (a single thread could be  responsible for a single partition) for example). 

Controlling number of files per partition allows controlling the size of single files which is important for performance.

**Comment by gijshendriksen** [comment]
I'd also be very happy to see this feature in DuckDB. We are using partitioned writes extensively in our project, and would love to have some more control over the size and number of the output Parquet files. A solution to this issue could also be to support the `FILE_SIZE_BYTES` Parquet output option for partitioned writes.

In case it's useful to show an actual use case where we run into issues with the partitioned Parquet writes, I'll explain our setup below, and how we want to use this DuckDB feature. I've also tried a bunch of workarounds, which I'll also share -- perhaps it can help someone else.
### Use case
For the [OpenWebSearch.eu](https://openwebsearch.eu/) project, we are crawling and indexing the Web to share with third parties. After crawling, we clean the HTML data and store the cleaned documents, along with a variety of metadata, as Parquet files in an S3 bucket. To make it easier to download the index files and corresponding metadata someone is looking for, we create index shards; currently based on the detected language of the document. As a result, we need to take the input Parquet files, partition them across the `language` column, and write them out to a different location in S3.
### Basic partitioned write approach
Initially, we tried to use the `FILE_SIZE_BYTES` option, to ensure files would have a reasonable size. The following is a simplified version of the  command we used:
```sql
COPY (FROM read_parquet('s3://ows/preprocessed/*.parquet'))
TO 's3://ows/shards/'
(FORMAT PARQUET, FILE_SIZE_BYTES '200mb', PARTITION_BY language);
```
Unfortunately, this gives the following error:
```
Error: Not implemented Error: Can't combine FILE_SIZE_BYTES and PARTITION_BY for COPY
```

An alternative would be to skip the `FILE_SIZE_BYTES` option:
```sql
COPY (FROM read_parquet('s3://ows/preprocessed/*.parquet'))
TO 's3://ows/shards/'
(FORMAT PARQUET, PARTITION_BY language);
```
But, as mentioned by @Hudler, this results in #num_cores * #num_partitions Parquet files, which is bad for performance.

We also tried @Hudler's suggestion of setting `threads = 1`, but this indeed makes the copying incredibly slow, and we would ideally want to make use of DuckDB's multi-threading capabilities.
### Using ORDER BY to reduce number of files
To solve the problem of every thread writing a Parquet file for every partition, we considered whether ordering the data by `language` would help. The intuition behind this approach was that ordering the data would cause rows with the same `language` column to be appointed to the same thread, limiting the total number of files being written.
```sql
COPY (SELECT * FROM read_parquet('s3://ows/preprocessed/*.parquet') ORDER BY language)
TO 's3://ows/shards/'
(FORMAT PARQUET, PARTITION_BY language);
```
This approach seemed to work: the total number of Parquet files being written decreased. However, there were still quite a few empty Parquet files being generated, i.e. files with only the Parquet header (we're unsure if this is a bug, or what could cause this behaviour). Removing all files smaller than 600 bytes resulted in a dataset with a reasonable amount of files.

However, this setup unfortunately no longer worked when scaling up to larger datasets. The `ORDER BY` became prohibitively expensive, and caused DuckDB to experience OOM errors by loading the full dataset into memory (even if specifying a local temporary directory for spilling to disk).
### Manual partitioning
After exhausting all pure DuckDB solutions we could think of, we also tried to manually perform the partitioning by iterating over all languages and writing the corresponding rows to individual directories (hoping that Parquet predicate pushdown would make this somewhat efficient). In Python, this would look like:
```python
import duckdb  
from tqdm import tqdm  

# DuckDB httpfs and S3 setup
...

duckdb.execute(f"SELECT DISTINCT language FROM 's3://ows/preprocessed/*.parquet'")  

for language, in tqdm(duckdb.fetchall()):  
    duckdb.execute(  
        "COPY (SELECT * EXCLUDE language FROM 's3://ows/preprocessed/*.parquet' WHERE language = ?) "  
        f"TO 's3://ows/shards/language={language}' "
        "(FORMAT PARQUET, FILE_SIZE_BYTES '200mb');",  
        [language]
    )
```
Unfortunately, this approach has to scan the full dataset once for every language, making it very slow -- even if predicate pushdown should make the filtering step more efficient.

Finally, we considered a two-step approach. First, we partition the data locally, using all cores, resulting in a large amount of small files. Then, we upload the individual directories back to S3, using `FILE_SIZE_BYTES`, to make the final files reasonably sized.
```python
import os

import duckdb
from tqdm import tqdm

# DuckDB httpfs and S3 setup
...

duckdb.execute(
    "COPY (FROM 's3://ows/preprocessed/*.parquet') "
    "TO '/tmp/preprocessed/' "
    "(FORMAT PARQUET, PARTITION_BY language);"
)

for subdir in tqdm(os.listdir('/tmp/preprocessed/')):
    duckdb.execute(
        "COPY (FROM read_parquet(concat(?, '/*.parquet'))) "
        f"TO 's3://ows/shards/{subdir}' "
        "(FORMAT PARQUET, FILE_SIZE_BYTES '200mb');",
        [f'/tmp/preprocessed/{subdir}']
    )
```

While still not ideal, this seems to work fine and is the workaround we're currently using. But adding support for controlling the file sizes / number of files in DuckDB's partitioned writes would be great, and would likely solve all our problems in this use case.

Of course, this setup of using DuckDB only to copy Parquet files from one location to another is likely not a core use case of DuckDB, but I imagine the same issues arise when exporting a DuckDB database to partitioned Parquet files.

**Comment by ylu3** [comment]
I don't think limiting the thread to 1 ensures one file per partition. I have encountered many cases where duckdb still generates multiple files if ORDER BY is present in query.
